src/ssl/PeekingPeerConnector.cc

```diff
@@ -338,8 +338,8 @@ Ssl::PeekingPeerConnector::noteNegotiationError(const Security::ErrorDetailPoint
     // thus hiding them.
     // Abort if no certificate found probably because of malformed or
     // unsupported server Hello message (TODO: make configurable).
-    if (!SSL_get_ex_data(session.get(), ssl_ex_index_ssl_error_detail) &&
-            (srvBio->bumpMode() == Ssl::bumpPeek  || srvBio->bumpMode() == Ssl::bumpStare) && srvBio->holdWrite()) {
+    if (!errorDetail->brokenCert() &&
```

Comment:
No bug is fixed. The old code would work too.
I just replaced the SSL_get_ex_data check with the simpler errorDetail->brokenCert() which in practice are equivalent.
The comment "Abort if no certificate found" refers to the check "if (serverCert)" inside the "if(! errorDetail->brokenCert() ..." check.

Reply:
> I just replaced the SSL_get_ex_data check with the simpler errorDetail->brokenCert() which in practice are equivalent.

What makes them equivalent? What prevents errorDetail from containing no broken certificate while ssl_ex_index_ssl_error_detail is set? I see no such invariants in the code. Yes, _today_, we only set ssl_ex_index_ssl_error_detail when we have a broken certificate, but that may change at any time! And, after we start checking for the broken certificate here, such a future change will not be connected to this code at all (but may break it; "may" because I am still not sure what this code is trying to accomplish and when).

```C++
/// the peer certificate (or nil)
Certificate *peerCert() { return peer_cert.get(); }

/// peer or intermediate certificate that failed validation (or nil)
Certificate *brokenCert() {return broken_cert.get(); }

void
Security::ErrorDetail::setPeerCertificate(const CertPointer &cert)
{
    assert(!broken_cert);
    peer_cert = cert;
    // unlike the constructor, the supplied certificate is not a broken_cert
}
```

As you can see, setPeerCertificate(), which we do call when detailing some errors, does not set broken_cert and even asserts that it was not set earlier.

We should not assume that ssl_ex_index_ssl_error_detail is only added when a broken certificate is found. If really necessary, we can add such an invariant by renaming ssl_ex_index_ssl_error_detail to something like ssl_ex_index_validation_error_detail_with_a_broken_certificate, but I hope that we do not have to do that.

> The comment "Abort if no certificate found" refers to the check `if (serverCert)` inside the `if (!errorDetail->brokenCert() ...` check.

Ouch. Well, that code will not be executed when there is no broken certificate. It sounds like that comment should be interpreted as "... but do _not_ abort if there is a broken certificate but no peer certificate" or something like that. The overall logic here is really unclear to me!

Perhaps we can avoid these problems by leaving this code as is? Is this change really in scope of this PR?

Reply:
> Is this change really in scope of this PR?

Not really.

But imagine the following:
We are peeking on server. We are sending to the remote TLS server the packets we received from the client, not an TLS client hello generated by squid. The squid/OpenSSL retrieves the Server Hello message but founds that it does not correspond to the TLS client hello generated localy. And produces an error.
If we are interesting to splice after peek then we can peek.
We only want to exclude cases the remote server certificate is not validate, or the remote server did not sent any certificate at all.
This is what the check does here. 

Because now we have update the passed error detail we can easily check if we had a broken (failed to verify) certificate.
We can leave unattached this code, however I think the change is more correct..

Reply:
```
ssl_bump peek all
ssl_bump splice all
```

I agree that the above Squid configuration should splice the connection with a TLS server that has a valid certificate but is slightly incompatible with Squid's TLS stack (e.g., has incompatible list of ciphers), resulting in a Security::Connect() error. With TLS v1.3 proliferation, such situations will probably become increasingly rare (because the certificate will be encrypted and, hence, unvalidatable by a splicing Squid), but they probably do exist for earlier TLS versions.

> We only want to exclude cases the remote server certificate is not validate, or the remote server did not sent any certificate at all.

Agreed (assuming that "exclude" means "ignore the error"). The accompanying documentation is very difficult for me to grok, but AFAICT, this code ignores Connect() error (and may eventually splice the connections) if "!brokenCert && serverCert".

> This is what the check does here.

IMO, "!brokenCert && serverCert" does not imply "valid" with _enough certainty_ for such a critical test. This is the crux of my concern -- this change substitutes one bad condition for another bad condition. Moreover, AFAICT, the new condition currently covers exactly the same cases than the old condition covers; it is just "shorter" or "simpler". We can and should fix this condition, but not in this PR because the correct fix is not trivial, and the problem itself is out of scope.

Most likely, a proper fix would be to provide a positive signal (e.g., a new "validated" flag or method of some sort) rather than rely on the lack of a _partial_ negative signal (i.e. "!broken && serverCert"). This should probably be done _after_ TLS v1.3 support is in to avoid redoing this change.

For now, let's undo this change.

If you think it is a good idea, please feel free to add an out-of-scope TODO comment:

```suggestion
    // TODO: Add/use a positive "successfully validated server cert" signal instead of
    // relying on the "![presumably_]validation_error && serverCert" signal combo.
    if (!SSL_get_ex_data(session.get(), ssl_ex_index_ssl_error_detail) &&
```

Reply:
commit 35559c8